sticker/src/encoder.rs

```diff
@@ -1,8 +1,22 @@
+use std::borrow::Borrow;
+
 use conllx::graph::{Node, Sentence};
 use failure::{format_err, Error};
 
 use crate::{Layer, LayerValue};
 
+/// Trait for sentence encoders.
+///
+/// A sentence encodes a representation of each token in a sentence,
+/// such as a part-of-speech tag or a topological field.
+pub trait SentenceDecoder {
+    type Encoding;
+
+    fn decode<E>(&self, labels: &[E], sentence: &mut Sentence) -> Result<(), Error>
+    where
+        E: Borrow<Self::Encoding>;
```

Comment:
Labels being a borrow of `Self::Encoding` seems somewhat counterintuitive and to my understanding requires an owner to borrow `E`. Currently there's the `Numberer` that provides the borrows which allows putting off allocation until the value is set on the token. There's not really much gained by putting off the allocation in my opinion.

I'm not sure if this would interfere with say binary classification where there's no numberer that provides borrows of the values.

Reply:
But in other decoders, not the whole struct is always used. E.g. in one of the dependency parser encoders, only one struct field is used, the other two fields are only used to determine the head.

> I'm not sure if this would interfere with say binary classification where there's no numberer that provides borrows of the values.

Note that decode also takes owned labels, they just need to borrowable (`Borrow` has an impl for any `T`).

Reply:
Wanted to add another thing to the first point: in the dependency decoders we'll actually want to return the n-best labels, so the encoding could actually be a `Vec` of the actual encoding. Here it pays of even more to borrow.